src/bin/pg_autoctl/pgsql.c

```diff
@@ -2562,17 +2641,35 @@ typedef struct IdentifySystem
  * contain the 'replication=1' parameter.
  */
 bool
-pgsql_identify_system(PGSQL *pgsql)
+pgsql_identify_system(PGSQL *pgsql, bool *unreachable)
 {
 	IdentifySystem context = { 0 };
 	char *sql = "IDENTIFY_SYSTEM";
 
 	PGconn *connection = NULL;
 	PGresult *result = NULL;
+	PGPing pingResult;
 
+	if (unreachable)
+	{
+		*unreachable = false;
+	}
 	connection = pgsql_open_connection(pgsql);
 	if (connection == NULL)
 	{
+		if (unreachable)
+		{
+			/*
+			 * Perform a PQping to determine why the connection open attempt was
+			 * unsuccessful.
+			 */
+			pingResult = PQping(pgsql->connectionString);
+			if (pingResult == PQPING_NO_RESPONSE)
+			{
+				*unreachable = true;
+			}
```

Comment:
Let's move the PQPing responsibility to the caller. And then think about unstable networks. See also the `PGConnStatus status` field in `PGSQL`, which might already solve your problem in a way that the caller can distinguish (as in https://github.com/citusdata/pg_auto_failover/blob/master/src/bin/pg_autoctl/service_keeper.c#L605 for instance):

``` c
/*
 * Allow higher level code to distinguish between failure to connect to the
 * target Postgres service and failure to run a query or obtain the expected
 * result. To that end we expose PQstatus() of the connection.
 *
 * We don't use the same enum values as in libpq because we want to have the
 * unknown value when we didn't try to connect yet.
 */
typedef enum
{
	PG_CONNECTION_UNKNOWN = 0,
	PG_CONNECTION_OK,
	PG_CONNECTION_BAD
} PGConnStatus;
```

Reply:
`PG_CONNECTION_BAD` is returned for both the "we couldn't make a network connection at all" and the "we connected but then your password was rejected" cases, unfortunately. `PQping()` differentiates between the two cases with `PQPING_NO_RESPONSE` and `PQPING_OK`.

Reply:
>think about unstable networks

In what way? The underlying connection code is already retrying several times. Or is there another angle you're referring to?

Reply:
>Let's move the PQPing responsibility to the caller

I think this forces us to duplicate most of the dance for formulating a replication conninfo string, but we may be able to get a good API out of it that can be used in both places. We'll give it a shot.

Reply:
Thanks for the explanation about needing PQPing to distinguish between the two cases. Something to the same effect should be added as a comment in the code. Unstable networks: failed to connect, then successful ping, because it's all up again. Maybe if we can ping we should retry connecting? I don't know.

About the API I guess we have the connection string in the caller, but I only say that now because I see that the change happens in pgsql.c and I don't remember that that's where we build connection strings.

Reply:
>Unstable networks: failed to connect, then successful ping, because it's all up again.

Yeah, this is IMO a deficiency in the libpq API. I want to know why a specific connection failed without having to issue a second connection, which might fail (or succeed) in a different way. But as far as I can tell, the current API does not give us that option.

Reply:
> About the API I guess we have the connection string in the caller, but I only say that now because I see that the change happens in pgsql.c and I don't remember that that's where we build connection strings.

The connection string is built inside `pgctl_identify_system` and is hidden from its callers. So unfortunately, we don't have access to the connection string in `cli_keeper_config_validate_replication` where we have to distinguish between the two cases of connection failure.

Reply:
I see now. Okay I think we should extend what we're doing in `pgsql_retry_open_connection` in order to keep the information we gather when using PGPing there. This retry mechanism is already being used for a connection of type PGSQL_CONN_UPSTREAM as used in `pgctl_identify_system`. 

So what about extending the `PGConnStatus` enum with some of the `PQPING_` diagnostic information and keeping track of what we know there? Or maybe we need a new enum for `PGConnPingStatus` and a new member in `struct PGSQL` to keep track of that?

That way we could differentiate between `PQPING_NO_RESPONSE` and `PQPING_OK` I suppose, without having to implement the PQPing retry mechanism again.